NEP-xxx: Tokenized Vault

edwardchew97 opened the pull request:
Reference implementations: https://github.com/Meteor-Wallet/tokenized-vault-nep-implementation

garikbesson commented:
Hi @edwardchew97, thank you for bringing new ideas and helping to push the entire ecosystem forward!

As a NEP moderator, I reviewed your proposal, and there is only one thing I would like to ask you - please, remove the `Unresolved Issues (Optional)` section from the proposal, since it's optional and hasn't been filled in.

Other than that, everything looks fine, and as soon as you push the requested changes, I will ask WG to assign SMEs to review the proposal.

garikbesson commented:
As a moderator, I'm asking @near/wg-contract-standards working group members to assign 2 Technical Reviewers to complete a technical review of the proposal.

IkerAlus reviewed (commented):
Hi @edwardchew97 thanks for the good work. I have done a general review of the proposal (not the implementation) and compared it with original [ERC4626 specification](https://eips.ethereum.org/EIPS/eip-4626), and I have a couple of comments:

- The proposal is missing `deposit` and `mint` methods. Is there any reason for this? AFAIK, these two methods are main state-changing flows in ERC4626.
- In ERC4626, `totalAssets` view method states that it MUST be inclusive of any fees that are charged against assets in the Vault. Can you add this statement too?
- Can you add the event emitted to the corresponding method description? For example, `withdraw` method should emit `withdraw` event.
- Same you did for `max_redeem`, with "Implementations should return `0` if...", can you add a similar statement for `maxWithdraw` method as done in ERC4626?
- Can you add when a method must or not revert (like it is done in the ERC4626)?
- Finally, in "Alternatives" it is mentioned this spec is not a "direct clone ERC-4626 clone without NEAR Adjustments".  Are this adjustments purely in the implementation? If they affect the spec, can you add a section stating which methods differ from  ERC-4626 spec and how?

edwardchew97 commented:
@IkerAlus Thanks a lot for the thorough review! Addressed all your points below:

> * The proposal is missing `deposit` and `mint` methods. Is there any reason for this? AFAIK, these two methods are main state-changing flows in ERC4626.
- Good catch — deposits/mints happen through ft_on_transfer. That was already in the implementation but not explicitly mentioned in the NEP. We’ve now added it to the spec for clarity.

> * In ERC4626, `totalAssets` view method states that it MUST be inclusive of any fees that are charged against assets in the Vault. Can you add this statement too?
- Added this to the spec, thanks!

> * Can you add the event emitted to the corresponding method description? For example, `withdraw` method should emit `withdraw` event.
- Great suggestion — events are now included in the method descriptions.

> * Same you did for `max_redeem`, with "Implementations should return `0` if...", can you add a similar statement for `maxWithdraw` method as done in ERC4626?
- Added that as well.

> * Can you add when a method must or not revert (like it is done in the ERC4626)?
- Updated the spec accordingly (no changes in the reference implementation).

> * Finally, in "Alternatives" it is mentioned this spec is not a "direct clone ERC-4626 clone without NEAR Adjustments".  Are this adjustments purely in the implementation? If they affect the spec, can you add a section stating which methods differ from ERC-4626 spec and how?
- Yes, most of the adjustments are only in the implementation and not shown on the interfaces. The differences are primarily due to the asynchronous nature of NEAR Protocol. This doesn’t necessarily prevent ERC-4626 from working but introduces some implementation-level nuances (e.g. `ft_on_transfer`). Beyond that, most choices are left to the developer, so we’ve opted to keep these adjustments in the implementation rather than the core spec. That said, we’ve added further clarification under the section. Appreciate you pointing this out too.

IkerAlus commented on neps/nep-0621.md:
`````diff
@@ -0,0 +1,491 @@
+````
+
+```rust
+pub trait VaultOptionalSlippageWithdraw {
`````
are these two new methods mandatory in the NEP or just recommended? 

I see they add a very valid use case but they are not part of ERC4626

IkerAlus commented on neps/nep-0621.md:
`````diff
@@ -0,0 +1,491 @@
+
+````rust
+pub trait FungibleTokenReceiver {
+    /// According to the NEP-141 standard, all fungible token transfers to a contract
`````
can we take out this explanation (from line 324 to 349) and put just below "Deposit and Mint" subsection?

IkerAlus commented on neps/nep-0621.md:
````diff
@@ -0,0 +1,491 @@
+}
+```
+
+Deposit and Mint
````
shouldn't this section be placed inside the "Contract Interface" section and not inside "Events" section?

IkerAlus commented on neps/nep-0621.md:
```diff
@@ -0,0 +1,491 @@
+    /// correctly handle `ft_on_transfer` according to NEP-141:
+    /// - Return `PromiseOrValue::Value(0)` if all tokens were accepted.
+    /// - Return a non-zero `U128` to indicate the number of tokens to refund.
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: U128, msg: String) -> PromiseOrValue<U128>
```
Both `deposit` and `mint` should emit a `VaultDeposit` event.  Something you indeed do in the ref impl in https://github.com/Meteor-Wallet/tokenized-vault-nep-implementation/blob/5952f914a4f1601aa73ac142d77af45685309933/src/lib.rs#L276 

This should be specified in the spec doc too for completeness.

IkerAlus reviewed (changes requested):
hey @edwardchew97 thanks for the thorough and detailed answers. I think the spec looks almost ready to go. I only have one main comment, the rest are format changes and clarifications.

fadeevab commented on neps/nep-0621.md:
```diff
@@ -0,0 +1,490 @@
+    /// Returns the total amount of underlying assets represented by all shares in existence.
+    ///
+    /// **Important:**
+    /// - Represents the vault's *total managed value*, not just assets held in the contract.
```
I understand where it comes from, but two "asset"/"assets" terms create a wild confusion. "Asset" means FT address, and "assets" means "amount of that one underlying asset (FT token)". Perhaps, it can be better addressed, e.g. "assets" could be called as "underlying asset's amount" in the docs comments.

fadeevab commented on neps/nep-0621.md:
```diff
@@ -0,0 +1,490 @@
+    /// Redeem a specific number of shares, **only if** they are worth
+    /// at least `min_amount` of assets.
+    /// This acts as a slippage-protected version of `redeem`.
+    fn redeem_with_slippage(
```
It seems both `redeem` and `withdraw` must have slippage control via `min_amount` parameter by default on NEAR, without "optional for security" `redeem/withdraw_with_slippage` counterparts. Because NEAR is fundamentally asynchronous in execution, and because this NEP is adaptation to NEAR's reality, it's worth to introduce secure by default methods, not vice versa.

fadeevab reviewed (commented):
Hi! I left two suggestions, check it out.

fadeevab reviewed (approved):
@gagdiez @edwardchew97 Sorry for delay, I missed the message. Besides of these 2 typos, it seems good to me, everything is fixed, and the long method names are very clear in meaning despite deviation from the original ERC :)

gagdiez commented on neps/nep-0621.md:
We can just note here that no major concerns were raised, besides the need to need to be careful with the `#security implications` sections described above
